lua/vsnip.lua

```diff
@@ -0,0 +1,182 @@
+end
+
+
+--- Wraps lines with quotes. @local
+-- @param lines array of strings
+-- @return array of lines, each wrapped with quotes.
+local wrap_lines = function (lines)
+  return vim.tbl_map(function(v)
+    return string.format([["%s"]], v) end,
+  lines)
+end
+
+
+
+--- Process prefix. @local
+-- @param prefix a string or strings operated by comma.
+-- @return array
+local process_prefix =  function(prefix)
+  if prefix:gmatch(",") then
+    return vim.fn.split(prefix, ",")
+  else
+    return {prefix}
+  end
+end
+
+
+
+--- Prompt user for an input. @local
+-- concatenate t & s and ask user for input.
+-- @param t & sstring
+-- @return string
+-- @todo use floating window :D
+local get_value = function(t, s)
+  return vim.fn.input(string.format("%s %s:> ", t, s))
+end
+
+
+
+--- Get a filetype's snippet filepath.
+-- Given a filetype, return its the file path and ensure it exists.
+-- @param ft string
+-- @return string: snippet filepath
+local get_snippet_file = function(ft)
+  local f = "json"
+  local file = ensure(string.format("%s/%s.%s", vim.g.vsnip_snippet_dir, ft, f))
+  return file
+end
+
+
+
+--- Create a new snippet table from opts or user input.
+-- when a opts is complete it return valid snippet table.
+-- @param opts table of `title`, `desc,` `prefix`, `body`
+-- @return table string to be inserted in
+M.create_snippet_entry = function(opts)
+  local opt = opts or {}
+  local title = opt.title or vim.fn.input("snippet title:> ")
+  return {
+    [title] = {
+      description = opt.desc or get_value(title, "snippet description"),
+      prefix = opt.prefix or process_prefix(get_value(title, "snippet prefix(es)")),
+      body = opt.body or get_value(title, "snippet body"),
+    }
+  }
+end
+
+
+
+--- Get snippets from filetype
+-- parses json file and returns lua table
+-- @param ft string
+-- @raise error vim when file can't be opend
+-- @return table of ft snippets
+M.get_snippets_for_ft = function(ft)
+  return vim.fn.json_decode(vim.fn.readfile(get_snippet_file(ft)))
+end
+
+
+
+--- Update ft's snippets.
+-- update filetype's snippet with new snippet
+-- @param ft string: representing the filetype.
+-- @param ns table: representing the new snippet to add.
+-- @todo get rid of jq dependency, prettify output without it!.
+-- @todo I don't like the fact it read, parse then write the file.
+-- find a way to just append to file without breaking json.
+M.update_snippets = function(ft, ns)
+  -- get current snippets table
+  local old = M.get_snippets_for_ft(vim.bo.filetype)
+  -- extend the old snippet with the new snippet
+  local str = vim.fn.json_encode(vim.tbl_extend("force", old, ns))
+  -- format with jq
+  local output = (function()
+    local tmp = vim.fn.tempname()
+    vim.fn.writefile({str}, tmp, "a")
+    local cmd = string.format("jq . %s", tmp)
+    return vim.fn.systemlist(cmd)
+  end)()
+  -- write output
+  local file = get_snippet_file(ft)
+  vim.fn.delete(file)
+  vim.fn.writefile(output, file, "a")
+end
```

Comment:
@Conni2461 I need your help with this one :smile: , basically this function takes a table, 
- concatenated with another table (which is the result of json file) and encode it,
- then runs a command to format json output and store the results in output,
- then deletes the file and then rewrite it again.

It sound very ugly, and I'd love to hear your thoughts. Thanks

Reply:
I think you can just overwrite the file by not adding a flag than you can get rid of the delete. 
I have to think about the other stuff give me a minute.

Reply:
Okay why is this local `output = (function()` even a thing you don't need a function here. which returns stuff. Just do `vim.fn.systemlist(cmd)`.
You also don't need to create a tmp file. You can pipe into jq. I think that should work somehow. see `jq --help`

Reply:
Wait a sec. Why can't you not just append the new stuff to the old file?

Reply:
> I think you can just overwrite the file by not adding a flag than you can get rid of the delete.

Nice, it worked, I was under the impression that a flag need to be supplied for the function to work, and that there is a flag for overwrite I don't know about. Thanks so much,

>  You don't need a function here which returns stuff

Oh this one was intentional to make it obvious where I'm getting the output from and what do I need to get it.  

> You also don't need to create a tmp file. You can pipe into jq. I think that should work somehow. see jq --help

This I spent sometime one but I always endup getting `zsh:1: parse error near `}'` thats when I do 

```lua
local cmd = string.format("jq . %s", str) -- "and echo %s | jq"
vim.fn.writefile(vim.fn.systemlist(cmd), get_snippet_file(ft))
```

Reply:
> Wait a sec. Why can't you not just append the new stuff to the old file? Because of possible duplications?

@Conni2461 that was my optimal solution that will let me skip a lot of s\**t .

However, the issue that I couldn't solve is appending the json string without breaking the json file.